src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,869 @@
+
+string_refinementt::string_refinementt(const namespacet &_ns, propt &_prop):
+  supert(_ns, _prop)
+{
+  use_counter_example=false;
+  variable_with_multiple_occurence_in_index=false;
+  initial_loop_bound=100;
+}
+
+void string_refinementt::set_mode()
+{
+  debug() << "initializing mode" << eom;
+  symbolt init=ns.lookup(irep_idt("__CPROVER_initialize"));
+  irep_idt mode=init.mode;
+  debug() << "mode detected as " << mode << eom;
+  generator.set_mode(mode);
+}
+
+void string_refinementt::display_index_set()
+{
+  for(auto i : index_set)
+  {
+    const exprt &s=i.first;
+    debug() << "IS(" << from_expr(s) << ")=={";
+
+    for(auto j : i.second)
+      debug() << from_expr(j) << "; ";
+    debug() << "}"  << eom;
+  }
+}
+
+// We compute the index set for all formulas, instantiate the formulas
+// with the found indexes, and add them as lemmas.
+void string_refinementt::add_instantiations()
+{
+  debug() << "string_constraint_generatort::add_instantiations: "
+          << "going through the current index set:" << eom;
+  for(std::map<exprt, expr_sett>::iterator i=current_index_set.begin(),
+         end=current_index_set.end(); i!=end; i++)
+  {
+    const exprt &s=i->first;
+    debug() << "IS(" << from_expr(s) << ")=={";
+
+    for(auto j : i->second)
+      debug() << from_expr(j) << "; ";
+    debug() << "}"  << eom;
+
+    for(auto j : i->second)
+    {
+      const exprt &val=j;
+
+      for(std::size_t k=0; k<universal_axioms.size(); k++)
+      {
+        exprt lemma=instantiate(universal_axioms[k], s, val);
+        add_lemma(lemma);
+      }
+    }
+  }
+}
+
+literalt string_refinementt::convert_rest(const exprt &expr)
+{
+  if(expr.id()==ID_function_application)
+  {
+    // can occur in __CPROVER_assume
+    bvt bv=convert_function_application(to_function_application_expr(expr));
+    assert(bv.size()==1);
+    return bv[0];
+  }
+  else
+  {
+    return supert::convert_rest(expr);
+  }
+}
+
+bvt string_refinementt::convert_symbol(const exprt &expr)
+{
+  const typet &type=expr.type();
+  const irep_idt &identifier=expr.get(ID_identifier);
+  assert(!identifier.empty());
+
+  if(refined_string_typet::is_unrefined_string_type(type))
+  {
+    string_exprt str=
+      generator.find_or_add_string_of_symbol(to_symbol_expr(expr));
+    bvt bv=convert_bv(str);
+    return bv;
+  }
+  else
+    return supert::convert_symbol(expr);
+}
+
+
+bvt string_refinementt::convert_function_application
+(const function_application_exprt &expr)
+{
+  debug() << "string_refinementt::convert_function_application "
+          << from_expr(expr) << eom;
+  exprt f=generator.add_axioms_for_function_application(expr);
+  return convert_bv(f);
+}
+
+bool string_refinementt::boolbv_set_equality_to_true(const equal_exprt &expr)
+{
+  if(!equality_propagation)
+    return true;
+
+  // We should not do that everytime, but I cannot find
+  // another good entry point
+  if(generator.get_mode()==ID_unknown)
+    set_mode();
```

Comment:
The constructor is called before the symbol table is filled so I could not find a way to figure out which mode to use at the point where the constructor is called. But maybe someone can suggest a better solution?